### PR TITLE
Enable EB notification and allow external accounts read only access

### PR DESCRIPTION
**Pull request by zumic96**
# Description

Enables EB notification by default so we can create rules.
Adds read only access for external accounts if any specified, if not it just creates default bucket policy as it was before the change.

Fixes [#1021](https://drivevariant.atlassian.net/browse/EAGER-1021)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Manual tests with and without `external_accounts` variable.

- [ ] Sanity Testing

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added documentation to test

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -123,3 +123,37 @@ resource "aws_s3_bucket_notification" "bucket_notification" {
 
   eventbridge = var.enable_bucket_notification
 }
+
+resource "aws_sns_topic" "sns" {
+  count = var.enable_bucket_notification && var.enable_sns_delivery ? 1 : 0
+  name  = aws_s3_bucket.bucket.id
+  tags  = var.tags
+
+  kms_master_key_id = "alias/aws/sns"
```
Comment (vibindaniel) [inline main.tf:132]: This key cannot be shared between accounts and hence a new key has to be created that has the ability to shared between accounts.

File: main.tf [hunk 2]
```diff
@@ -123,3 +123,37 @@ resource "aws_s3_bucket_notification" "bucket_notification" {
...
+}
+
+module "eb_rule" {
```
[the remark this answers is not in the document]
Comment (zumic96) [reply main.tf:135]: added an example for SQS
[example](https://github.com/variant-inc/terraform-aws-s3/pull/11/files#diff-a1abad2da4f74766841d88a2acfcdf090876bcec52b4e9cf1b8779443a6fb30fR12-R14)

File: examples/event_notification/main.tf [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  bucket_prefix              = "usx-store"
+  enable_bucket_notification = true
+  enable_sns_delivery        = true
+  
+  event_targets = {
+    target-lambda-test = {
+      arn = "arn:aws:lambda:us-east-1:123456789012:function:lambda"
+    }
+    target-sqs-test = {
+      arn = "arn:aws:sqs:us-east-1:123456789012:sqs"
```
Comment (vibindaniel) [inline examples/event_notification/main.tf:13]: Every time a project needs access to the eb through sqs, do they have to request out to you?

File: main.tf (hunk 1 above)
Comment (vibindaniel) [inline main.tf:132]: Was this tested anywhere?

Comment (zumic96) [reply main.tf:132]: Initially I tested creating SNS topic but not the full flow
As I had test setup for cross account EB buses, I've added SNS target and it indeed didn't work with AWS managed CMK, had to create new one and set specific policies for it.
I'll add creation of it and setting minimal policies with option to add custom ones for cross-account access.

**Comment by zumic96** [comment]
Updated and tested on data-dev account, didn't delete for checkup purposes
Resources:
[S3](https://s3.console.aws.amazon.com/s3/buckets/test-luka-20220407090348255300000001?region=us-east-1&tab=objects)
[EB rule](https://us-east-1.console.aws.amazon.com/events/home?region=us-east-1#/eventbus/default/rules/test-luka--eb-rule)
[IAM Role](https://us-east-1.console.aws.amazon.com/iamv2/home?region=us-east-1#/roles/details/EventBridge-rule-test-luka--eb-rule?section=permissions)
[SNS Topic](https://us-east-1.console.aws.amazon.com/sns/v3/home?region=us-east-1#/topic/arn:aws:sns:us-east-1:854762601885:test-luka-20220407090348255300000001)
[KMS Key](https://us-east-1.console.aws.amazon.com/kms/home?region=us-east-1#/kms/keys/d8278083-8768-467f-a483-9da00ebfed44)
[SQS Queue](https://us-east-1.console.aws.amazon.com/sqs/v2/home?region=us-east-1#/queues/https%3A%2F%2Fsqs.us-east-1.amazonaws.com%2F854762601885%2Fluka-delete-0405)
Subscription to SNS was done manually, as I guess this will vary depending on the use case.

File: main.tf [hunk 4]
```diff
@@ -123,3 +125,181 @@ resource "aws_s3_bucket_notification" "bucket_notification" {
...
+
+  policy_id = "kms_policy"
+
+  statement {
+    sid    = "Allow access for Key Administrators"
+    effect = "Allow"
+    principals {
+      type = "AWS"
+      identifiers = [
+        "arn:aws:iam::854762601885:role/aws-reserved/sso.amazonaws.com/AWSReservedSSO_AWSAdministratorAccess_d03a558f68b16416"
```
Comment (vibindaniel) [inline main.tf:238]: Why is this hardcoded?

File: main.tf [hunk 5]
```diff
@@ -123,3 +125,181 @@ resource "aws_s3_bucket_notification" "bucket_notification" {
 
   eventbridge = var.enable_bucket_notification
 }
+
+resource "aws_sns_topic" "sns" {
```
Comment (vibindaniel) [inline main.tf:129]: SNS & KMS should be in its own module. This is adding lots of scope into the s3 module, which is for creating s3 buckets.

**Review by vibindaniel — CHANGES_REQUESTED** [review]

File: main.tf (hunk 5 above)
Comment (zumic96) [reply main.tf:129]: I agree, do you think we should still use those modules inside s3 module and that way hide this configuration from end user, similarly as it's being done with EB rule or just do it in final code where we'll also use s3 module?
What would be more in accordance with internal code practices?

Comment (zumic96) [reply main.tf:129]: @vibindaniel let me know which approach is preferred

**Comment by zumic96** [comment]
@vibindaniel I removed all other stuff and just made bucket event notifications to EventBridge bus as default.

Wasn't sure about the bucket policies so didn't touch those, should I make data account (648462982672) default for all buckets or should we add some kind of input? Especially if we'll want some access from data-dev in future or some other account.

**Comment by vibindaniel** [comment]
> @vibindaniel I removed all other stuff and just made bucket event notifications to EventBridge bus as default.
> 
> Wasn't sure about the bucket policies so didn't touch those, should I make data account (648462982672) default for all buckets or should we add some kind of input? Especially if we'll want some access from data-dev in future or some other account.

The target account needs to be an input variable.

**Comment by vibindaniel** [comment]
@zumic96 Can we have a call again tomorrow at the same time as yesterday?

**Comment by zumic96** [comment]
> @zumic96 Can we have a call again tomorrow at the same time as yesterday?

Sure thing.

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -1,5 +1,5 @@
 repos:
-  - repo: git://github.com/antonbabenko/pre-commit-terraform
+  - repo: https://github.com/antonbabenko/pre-commit-terraform
     rev: v1.51.0
```
Comment (naveen-variant) [inline .pre-commit-config.yaml:3]: Can we also change this to latest version v1.68.1

**Review by vibindaniel — APPROVED** [review]

File: .pre-commit-config.yaml (hunk 6 above)
Comment (zumic96) [reply .pre-commit-config.yaml]: updated to latest version, could you please reapprove?

**Review by jahnavimaddi — COMMENTED** [review]
Created a test bucket by using this module in dpl and verified read-only access for the external accounts mentioned and the event bucket notification is enabled. Please refer below screenshots.

Please squash the commits before merge and address all the unresolved threads.
Screen Shot 2022-04-27 at 10 54 57 AM
Screen Shot 2022-04-27 at 11 00 11 AM

**Comment by naveen-variant** [comment]
Created bucket, i see event bridge is enabled and read only access is enabled for external account.

**Review by naveen-variant — APPROVED** [review]

**Review by jahnavimaddi — APPROVED** [review]

**Pull request merged**